### DISCUSSION TITLE
Any plans for implementing aggregation pushdown?

### DISCUSSION THREAD

**Discussion by yingfeng** [duckdb/duckdb#4515] [General]
It seems agg pushdown for join has not been implemented, which means join over non-primary keys will meet large intermediate results, or performance could have been improved.
Does there exist any plan for such kinds of optimization? 

```
Push down Join
Before:
        Join
      /      \
  LEFT     Aggregation
               \
                ....


After:
     Aggregation
          |
         Join
        /    \
     LEFT   ....

```

**Comment by hannes** [comment]
Happy to review a PR

**Comment by hawkfish** [comment]
Does this relate to GroupJoin?

**Reply by lnkuiper** [reply]
GroupJoin is somewhat related, but this is a separate optimization. Eager aggregation can reduce intermediates early. There are queries/data sets where this optimization is worthwhile. I spoke to someone who implemented this optimization in a different system, and it sounded like it was only good in niche cases.
